### PR TITLE
Ss at filter list

### LINKED ISSUE #9: 9. As a user, I want to filter my shopping list to make it easier to locate an item in the list.
AC:

* Display a text field above the top of the shopping list
* As the user types into the field, the list should narrow to display only items that contain the text the user entered in the filter field
* When the field has text in it, the user should be able to tap a UI element (e.g., with an "X" button next to the field) to clear the field
* The filter text should match any part of the item name (i.e. it should _not_ only match from the start of the string)

### PR DESCRIPTION
**Pull request by audracodes**
## Description

A search bar is displayed above the user's list which allows them to search their items. The list updates as the user types, and the search bar can be cleared by clicking the clear button. The filtered text will match any part of the item name, not just the beginning characters.

## Related Issue

Closes #9 

## Acceptance Criteria

- [x] Display a text field above the top of the shopping list

- [x] As the user types into the field, the list should narrow to display only items that contain the text the user entered in the filter field

- [x] When the field has text in it, the user should be able to tap a UI element (e.g., with an "X" button next to the field) to clear the field

- [x] The filter text should match any part of the item name (i.e. it should not only match from the start of the string)

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

Added `filterValue` and `filteredList` to state. `filteredList` uses the `userList` in Context, so the database does not need to be called constantly. Added a search input above the list, typing in the search bar updates `filterValue`, filters through the list items, and then updates `filteredList`. Added a button which sets `filterValue` to an empty string when clicked in order to clear the search input.Changed what is mapped to the printed list from the list in context to  `filteredList`. The list is filtered based on `filterValue` before printing, and updates as the user types into the search bar.

### Before

Screen Shot 2020-11-05 at 1 26 28 PM

### After

Screen Shot 2020-11-05 at 1 27 02 PM

## Testing Steps / QA Criteria

On opening the ViewList component, there should be a search bar and clear button above the list. Open the ViewList component in your browser's dev tools. The second item in state should update in real time for whatever is typed into the search bar (`filterValue`). The third item in state, which are the items in the database, should filter based on was is typed into the search bar. The app should only render items that match the characters in the search bar, regardless on where they appear in the word (ex. Typing 'oo' will still show 'food'). The clear button should reset the search input both on the app and in state.

### CODE REVIEW

**Comment by rbompiani** [comment]
Nicely done!  Very clean and easy to understand.  Looks like all the AC is taken care of, and everything works as expected.

These items weren't included in the AC, but what do you guys think about adding these as future issues (if we have time):

1. when the search returns no items, display a message that no items match the search criteria
2. Should we include some of the logic from the function in AddItems that checks for duplicates in the database? For example, if searching "ice cream," "ice-cream" or "icecream" would ignore spaces and punctuation and return all the variants.

I'm not totally sure about 2, it could create some confusion for the user where they type "-" and nothing happens, making it seem like a bug. And, theoretically, by the time you've typed a few characters you should have a small enough filtered list that finding your item isn't an issue.

File: src/components/ViewList.js [hunk 1]
```diff
@@ -73,12 +82,36 @@ const ViewList = () => {
...
+    setFilterValue(e.target.value);
+  };
+
+  useEffect(() => {
+    let listFilter = currentList.userList;
+    let filtered =
+      listFilter &&
+      listFilter.filter((item) => {
+        return item.itemName.toLowerCase().includes(filterValue.toLowerCase());
+      });
```
Comment (rbompiani) [inline src/components/ViewList.js:95]: This is a thing of beauty

**Review by rbompiani — APPROVED** [review]

**Comment by audracodes** [comment]
> Nicely done! Very clean and easy to understand. Looks like all the AC is taken care of, and everything works as expected.
> 
> These items weren't included in the AC, but what do you guys think about adding these as future issues (if we have time):
> 
>     1. when the search returns no items, display a message that no items match the search criteria
> 
>     2. Should we include some of the logic from the function in AddItems that checks for duplicates in the database? For example, if searching "ice cream," "ice-cream" or "icecream" would ignore spaces and punctuation and return all the variants.
> 
> 
> I'm not totally sure about 2, it could create some confusion for the user where they type "-" and nothing happens, making it seem like a bug. And, theoretically, by the time you've typed a few characters you should have a small enough filtered list that finding your item isn't an issue.

Thank you very much! I think the first one would be a great issue to add. I'm not entirely sure about the second one either, but it's worth asking the mentors about!

**Review by AliLynne — APPROVED** [review]
This is neat and clean and easy to follow. Great job at naming variables and functions for your filter and filtered lists.

**Pull request merged**